README.md
# User Service

A flask server to manage users and authentication


## Running
``` bash
docker pull ghcr.io/inquisitev/user-service:main
```
``` bash
docker run -p 3004:3004 ghcr.io/inquisitev/user-service:main
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.1.0] - 2023-03-05

### Added

- Create user
- Log in as user and get token
- verify token for ueser
- get user info

src/test_AuthService.py
import pytest
from src.authService import app, persister


# must clear the user map so that adding the same test user multiple times doesnt break
# the test suite
@pytest.fixture(autouse=True)
def reset_test_env():
    persister.user_map.clear()


# test that i can add a user and verify with the auth token i am given
def test_adding_a_user():
    client = app.test_client()

    response = client.post(
        "/users/new",
        json={
            "first_name": "duckey",
            "last_name": "beaver",
            "email": "duckey@olivesToys.com",
            "password": "mollyAndOliveAreBuddies",
        },
    )

    required_keys = ["expires_on", "token"]
    assert all([x in response.json for x in required_keys])
    token = response.json["token"]
    assert response.status_code == 200

    response = client.post(
        "/users/verify", json={"token": token, "email": "duckey@olivesToys.com"}
    )

    assert response.status_code == 200
    assert response.json["verified"] is True


# test that a wrong token will not verify
def test_a_incorrect_token_will_not_verify():
    client = app.test_client()

    response = client.post(
        "/users/new",
        json={
            "first_name": "duckey",
            "last_name": "beaver",
            "email": "chickum@olivesToys.com",
            "password": "mollyAndOliveAreBuddies",
        },
    )

    required_keys = ["expires_on", "token"]
    assert all([x in response.json for x in required_keys])
    assert response.status_code == 200

    response = client.post(
        "/users/verify",
        json={"token": "INVALID_TOKEN", "email": "duckey@olivesToys.com"},
    )

    assert response.status_code == 200
    assert response.json["verified"] is False


# test that a wrong token will not verify
def test_a_different_users_token_will_not_verify():
    client = app.test_client()

    response = client.post(
        "/users/new",
        json={
            "first_name": "duckey",
            "last_name": "beaver",
            "email": "chickum@olivesToys.com",
            "password": "mollyAndOliveAreBuddies",
        },
    )

    token = response.json["token"]

    response = client.post(
        "/users/new",
        json={
            "first_name": "duckey",
            "last_name": "beaver2",
            "email": "beaver@olivesToys.com",
            "password": "mollyAndOliveAreBuddies",
        },
    )

    response = client.post(
        "/users/verify", json={"token": token, "email": "beaver@olivesToys.com"}
    )

    assert response.status_code == 200
    assert response.json["verified"] is False


# test that when adding a new user, if the email is already registered, then do not
# provide auth token
def test_emails_must_be_unique():
    client = app.test_client()

    response = client.post(
        "/users/new",
        json={
            "first_name": "duckey",
            "last_name": "beaver",
            "email": "chickum@olivesToys.com",
            "password": "mollyAndOliveAreBuddies",
        },
    )

    required_keys = ["expires_on", "token"]
    assert all([x in response.json for x in required_keys])
    assert response.status_code == 200

    response = client.post(
        "/users/new",
        json={
            "first_name": "Daddy",
            "last_name": "Dolphin",
            "email": "chickum@olivesToys.com",
            "password": "mollyAndOliveAreBuddies",
        },
    )

    assert response.status_code == 409
    required_keys = ["expires_on", "token"]
    assert not any([x in response.json for x in required_keys])
    assert "message" in response.json
    assert response.json["message"] == "An account with that email already exists."


# test getting the details for a user


def test_a_user_can_log_in_if_they_haven_an_account():
    client = app.test_client()

    response = client.post(
        "/users/new",
        json={
            "first_name": "duckey",
            "last_name": "beaver",
            "email": "duckey@olivesToys.com",
            "password": "mollyAndOliveAreBuddies",
        },
    )

    response = client.post(
        "/users/login",
        json={"email": "duckey@olivesToys.com", "password": "mollyAndOliveAreBuddies"},
    )

    assert response.status_code == 200
    required_keys = ["expires_on", "token"]
    assert all([x in response.json for x in required_keys])


def test_a_user_can_not_log_in_if_they_dont_have_an_account():
    client = app.test_client()

    response = client.post(
        "/users/login",
        json={"email": "duckey@olivesToys.com", "password": "mollyAndOliveAreBuddies"},
    )

    assert response.status_code == 409
    assert "message" in response.json
    assert response.json["message"] == "There is no account with that email."


def test_a_user_can_not_log_in_if_they_have_the_wrong_password():
    client = app.test_client()

    response = client.post(
        "/users/new",
        json={
            "first_name": "duckey",
            "last_name": "beaver",
            "email": "duckey@olivesToys.com",
            "password": "mollyAndOliveNOTAreBuddies",
        },
    )

    response = client.post(
        "/users/login",
        json={"email": "duckey@olivesToys.com", "password": "mollyAndOliveAreBuddies"},
    )

    assert response.status_code == 409
    assert "message" in response.json
    assert response.json["message"] == "Incorrect credentials."


def test_get_into_of_a_user_that_does_not_exist():
    client = app.test_client()

    response = client.post(
        "/users/user_info",
        json={"email": "duckey@olivesToys.com", "token": "982034576_fgasdfa"},
    )

    assert response.status_code == 409
    assert "message" in response.json
    assert response.json["message"] == "There is no account with that email."


def test_get_info_of_a_user_that_exists():
    client = app.test_client()

    client.post(
        "/users/new",
        json={
            "first_name": "duckey",
            "last_name": "beaver",
            "email": "duckey@olivesToys.com",
            "password": "mollyAndOliveNOTAreBuddies",
        },
    )

    response = client.post(
        "/users/login",
        json={
            "email": "duckey@olivesToys.com",
            "password": "mollyAndOliveNOTAreBuddies",
        },
    )

    token = response.json["token"]

    response = client.post(
        "/users/user_info", json={"email": "duckey@olivesToys.com", "token": token}
    )

    assert response.status_code == 200
    assert response.json == {
        "first_name": "duckey",
        "last_name": "beaver",
        "email": "duckey@olivesToys.com",
    }


def test_get_info_for_user_that_exists_but_authentication_fails():
    client = app.test_client()

    client.post(
        "/users/new",
        json={
            "first_name": "duckey",
            "last_name": "beaver",
            "email": "duckey@olivesToys.com",
            "password": "mollyAndOliveNOTAreBuddies",
        },
    )

    response = client.post(
        "/users/user_info",
        json={"email": "duckey@olivesToys.com", "token": "INCORRECT_TOKEN"},
    )

    assert response.status_code == 409
    assert response.json == {
        "message": "Incorrect credentials.",
    }
    

def test_index():
    client = app.test_client()

    response = client.get(
        "/",
    )
    assert response.status_code == 200
    assert response.json == {
    }



src/authService.py
from flask import Flask, request
from datetime import date, timedelta
from dataclasses import dataclass
from typing import List, Tuple
import hashlib
import hmac
import os
import secrets
import string

class Cryptographer:
    def __init__(self, secret_key):
        self.secret_key = secret_key

    def secure_pw(self, plain_text: str) -> Tuple[bytes, bytes]:
        salt = os.urandom(16)
        pw_hash = hashlib.pbkdf2_hmac('sha256', plain_text.encode(), salt, 100000)
        return salt, pw_hash
    
    def check_pw(self, plain_text, hashed_pw, salt):
        return hmac.compare_digest(
            hashed_pw,
            hashlib.pbkdf2_hmac('sha256', plain_text.encode(), salt, 100000)
        )
        
def generate_api_key(length=32):
    alphabet = string.ascii_letters + string.digits
    api_key = ''.join(secrets.choice(alphabet) for _ in range(length))
    return api_key



@dataclass
class User:
    first_name: str
    last_name: str
    email: str
    encrypted_pass: bytes
    salt: bytes

    def to_json(self):
        return {
            "first_name": self.first_name,
            "last_name": self.last_name,
            "email": self.email,
        }


@dataclass
class Token:
    token: str
    expires_on: date

    def to_json(self):
        return {"token": self.token, "expires_on": self.expires_on}


class AuthenticationTokenHandler:
    def generate_token(self, user: User) -> Token:
        # TODO: Actually make a token
        return Token(generate_api_key(), date.today() + timedelta(days=180))


class UserPerstance:
    # TODO: Actually store things in a database
    # TODO: secure_pw data before it is put into the database and decrypt it after

    def __init__(self):
        self.user_map = {}

    def user_is_known(self, email: str) -> bool:
        """Check if the user is known in this persistent storage apparatus

        Args:
            email (str): the email of the user

        Returns:
            bool: true if the user is known
        """
        return email in self.user_map

    def get_user(self, email: str) -> User:
        """get the user with the associated email, if possible

        Args:
            email (str): the email of the user

        Returns:
            User: the user with the email address, if theyre known.
        """
        if self.user_is_known(email):
            user_struct = self.user_map[email]
            user = user_struct["user"]
            return user

        return None

    def get_tokens(self, email: str) -> List[Token]:
        """get all of the tokens for the user with the email address

        Args:
            email (str): the email of the user

        Returns:
            List[Token]: the tokens registered to the user
        """
        if self.user_is_known(email):
            user_struct = self.user_map[email]
            tokens = user_struct["tokens"]
            return tokens

        return []

    def add_token_to_user(self, email: str, token: str) -> bool:
        """Save a token to the user with email

        Args:
            email (str): the email of the user
            token (str): the token to save

        Returns:
            bool: true if successfully saved
        """
        user = self.get_user(email)
        if user is not None:
            self.user_map[email]["tokens"].append(token)
            return True

    def add_user(self, user: User) -> bool:
        """add a user to the persistent storage apparatus

        Args:
            user (User): the user to save

        Returns:
            bool: true if successfully saved
        """
        if not self.user_is_known(user.email):
            self.user_map[user.email] = {"user": user, "tokens": []}
            return True
        return False


class UserService:
    def __init__(self, crypto, token_factory, persistance):
        self.crypto = crypto
        self.token_factory = token_factory
        self.persistance = persistance

    def create_user(
        self, first_name: str, last_name: str, email: str, password: str
    ) -> User:
        """Create a user

        Args:
            first_name (str): The first name of the user
            last_name (str): The last name of the user
            email (str): The email of the user
            password (str): The password of the user

        Returns:
            User: The details of the user that has been added, excluding password.
            if the user has not been added for some reason, return None
        """
        salt, encrypted_pass = crypto.secure_pw(password)
        user = User(first_name, last_name, email, encrypted_pass, salt)
        if self.persistance.add_user(user):
            return user
        else:
            return None

    def get_user(self, email: str) -> User:
        """Get the user with the email from persistent storage

        Args:
            email (str): the email of the user

        Returns:
            User: the User with the email address, None if 
            retrieved from persistent storage as None
        """
        return self.persistance.get_user(email)

    def user_does_exist(self, email: str) -> bool:
        """Check if the user with email exists

        Args:
            email (str): the email of the user

        Returns:
            bool: true if the user exists
        """
        return self.persistance.get_user(email) is not None

    def authenticate_user(self, email: str, password: str) -> Token:
        """Generate a new token for a user provided that
        they have supplied correct credentials

        Args:
            email (str): the email of the user
            password (str): the plaintext password that will
                be verified against the stored encrypted password

        Returns:
            Token: The new authentication token attached to the user
                with the email
        """
        user = self.persistance.get_user(email)
        if user is not None:
            pw_is_correct = crypto.check_pw(password, user.encrypted_pass, user.salt)
            if pw_is_correct:
                new_token = self.get_new_token(user)
                return new_token

        return None

    def get_new_token(self, user: User) -> Token:
        """Get a new token for the user

        Args:
            user (User): the user to create a new token for.

        Returns:
            Token: the new token used for authentication. 
        """
        token = self.token_factory.generate_token(user)
        self.persistance.add_token_to_user(user.email, token)
        return token

    def verify_token(self, email: str, token: Token) -> bool:
        """Verify an authenitcation token with what is known already

        Args:
            email (str): the email of the user
            token (str): the token to attempt verification with

        Returns:
            bool: true if verification successful
        """
        tokens = self.persistance.get_tokens(email)
        return any([token == ctoken.token for ctoken in tokens])


app = Flask(__name__)

persister = UserPerstance()
crypto = Cryptographer("SECRET_TUNNEL")
authenticator = AuthenticationTokenHandler()
user_service = UserService(crypto, authenticator, persister)


@app.route("/", methods=["GET"])
def index():
    return {}, 200

@app.route("/users/new", methods=["POST"])
def create_user():
    request_data = request.json

    user = user_service.create_user(**request_data)
    if user is not None:
        token = user_service.get_new_token(user)
        response_payload = token.to_json()
        return response_payload, 200

    else:
        return {"message": "An account with that email already exists."}, 409


@app.route("/users/verify", methods=["POST"])
def verify_user():
    request_data = request.json
    verified = user_service.verify_token(**request_data)
    response_payload = {"verified": verified}

    return response_payload, 200


@app.route("/users/login", methods=["POST"])
def authenticate_user():
    request_data = request.json
    if not user_service.user_does_exist(request_data["email"]):
        message = {"message": "There is no account with that email."}
        return message, 409
    else:
        new_token = user_service.authenticate_user(**request_data)

        if new_token is not None:
            return new_token.to_json(), 200
        else:
            message = {"message": "Incorrect credentials."}
            return message, 409


@app.route("/users/user_info", methods=["POST"])
def lookup_user():
    request_data = request.json
    email = request_data["email"]
    if not user_service.user_does_exist(email):
        message = {"message": "There is no account with that email."}
        return message, 409
    else:
        verified = user_service.verify_token(**request_data)

        if verified:
            user = user_service.get_user(email)
            return user.to_json(), 200
        else:
            message = {"message": "Incorrect credentials."}
            return message, 409
